General Timer Improvements

LubricantJam opened the pull request:
- Changed the way checkpoints work, they are no longer field dependent and can be scalable. (For maps like surf_classics etc)
- Velocity saving for leaving and entering a zone allowing players to improve their pre-speed. 
- Added a minor fix for zone groups not loading on the startup map (rare instance).
- Fixed the issue where silent spec wouldn't remove you from the players spec list, only from your own. (Fixes #39)
- Added a convar to disable VIP commands as some servers have external VIP modules which may conflict.
- Fixed issue where scoreboard would display Unranked players at the top of the scoreboard (if they had not yet been inserted into ck_playerrank) (Fixes #59).

Nothing is 100%, I have tested and running the smx on my server right now and all seems to work fine, it'll just need to be tidied up if need be and minor improvements as this is from the old private fork.

qawery-just-sad commented:
Tested by upgrading from master 284 to your dev
Client "Just Sad" connected (xxx.xx.x.x:27015).
L 04/25/2020 - 10:21:13: [surftimer.smx] [Surftimer] SQL Error (sql_selectMapRecordCallback): Unknown column 'velStartXY' in 'field list'
L 04/25/2020 - 10:21:13: [surftimer.smx] [Surftimer] SQL Error (SQL_selectFastestBonusCallback): Unknown column 'velStartXY' in 'field list'
L 04/25/2020 - 10:21:13: [surftimer.smx] [Surftimer] SQL Error (sql_selectRecordCheckpointsCallback): Unknown column 'cp' in 'field list'
L 04/25/2020 - 10:21:14: [surftimer.smx] [Surftimer] SQL Error (sql_viewStageRecordsCallback): Unknown column 'velStartXY' in 'field list'
L 04/25/2020 - 10:21:14: [SurfTimer] Finished loading server settings in 0.644531s
L 04/25/2020 - 10:21:16: [Surftimer] Loading STEAM_1:0:6xxxxx - Just Sad settings
L 04/25/2020 - 10:21:16: [surftimer.smx] [Surftimer] SQL Error (SQL_selectPersonalRecordsCallback): Unknown column 'velStartXY' in 'field list'
L 04/25/2020 - 10:21:16: [surftimer.smx] [Surftimer] SQL Error (SQL_selectPersonalBonusRecordsCallback): Unknown column 'velStartXY' in 'field list'
L 04/25/2020 - 10:21:16: [surftimer.smx] [Surftimer] SQL Error (SQL_selectPersonalStageRecordsCallback): Unknown column 'velStartXY' in 'field list'
L 04/25/2020 - 10:21:16: [SurfTimer] STEAM_1:0:66814046: Finished db_viewPlayerPoints in 0.185546s
L 04/25/2020 - 10:21:16: [SurfTimer] STEAM_1:0:66xxxxx: Finished db_GetPlayerRank in 0.214843s
L 04/25/2020 - 10:21:16: [SurfTimer] STEAM_1:0:6xxxxxx: Finished db_viewPlayerOptions in 0.019531s
L 04/25/2020 - 10:21:16: [SurfTimer] STEAM_1:0:6xxxxxx: Finished db_CheckVIPAdmin in 0.019531s
L 04/25/2020 - 10:21:16: [SurfTimer] STEAM_1:0:6xxxxxx Finished db_viewCustomTitles in 0.019531s
L 04/25/2020 - 10:21:16: [surftimer.smx] [SurfTimer] SQL Error (SQL_selectCheckpointsCallback): Unknown column 'cp' in 'field list'

guess we'll need some upgrade script for dbs

samatazz commented:
Yeah upgrade scripts will be needed. I am curious as to performance in larger checkpoint maps and also why 4 different fields for a velocity? I get you want some stats but is the entry velocity not good enough for checkpoints? Putting all that extra data seems slightly overkill especially for servers with thousands of completes on maps with higher checkpoints

LubricantJam commented:
> Tested by upgrading from master 284 to your dev
> Client "Just Sad" connected (xxx.xx.x.x:27015).
> L 04/25/2020 - 10:21:13: [surftimer.smx] [Surftimer] SQL Error (sql_selectMapRecordCallback): Unknown column 'velStartXY' in 'field list'
> L 04/25/2020 - 10:21:13: [surftimer.smx] [Surftimer] SQL Error (SQL_selectFastestBonusCallback): Unknown column 'velStartXY' in 'field list'
> L 04/25/2020 - 10:21:13: [surftimer.smx] [Surftimer] SQL Error (sql_selectRecordCheckpointsCallback): Unknown column 'cp' in 'field list'
> L 04/25/2020 - 10:21:14: [surftimer.smx] [Surftimer] SQL Error (sql_viewStageRecordsCallback): Unknown column 'velStartXY' in 'field list'
> L 04/25/2020 - 10:21:14: [SurfTimer] Finished loading server settings in 0.644531s
> L 04/25/2020 - 10:21:16: [Surftimer] Loading STEAM_1:0:6xxxxx - Just Sad settings
> L 04/25/2020 - 10:21:16: [surftimer.smx] [Surftimer] SQL Error (SQL_selectPersonalRecordsCallback): Unknown column 'velStartXY' in 'field list'
> L 04/25/2020 - 10:21:16: [surftimer.smx] [Surftimer] SQL Error (SQL_selectPersonalBonusRecordsCallback): Unknown column 'velStartXY' in 'field list'
> L 04/25/2020 - 10:21:16: [surftimer.smx] [Surftimer] SQL Error (SQL_selectPersonalStageRecordsCallback): Unknown column 'velStartXY' in 'field list'
> L 04/25/2020 - 10:21:16: [SurfTimer] STEAM_1:0:66814046: Finished db_viewPlayerPoints in 0.185546s
> L 04/25/2020 - 10:21:16: [SurfTimer] STEAM_1:0:66xxxxx: Finished db_GetPlayerRank in 0.214843s
> L 04/25/2020 - 10:21:16: [SurfTimer] STEAM_1:0:6xxxxxx: Finished db_viewPlayerOptions in 0.019531s
> L 04/25/2020 - 10:21:16: [SurfTimer] STEAM_1:0:6xxxxxx: Finished db_CheckVIPAdmin in 0.019531s
> L 04/25/2020 - 10:21:16: [SurfTimer] STEAM_1:0:6xxxxxx Finished db_viewCustomTitles in 0.019531s
> L 04/25/2020 - 10:21:16: [surftimer.smx] [SurfTimer] SQL Error (SQL_selectCheckpointsCallback): Unknown column 'cp' in 'field list'
> 
> guess we'll need some upgrade script for dbs

No, you shouldn't need them at all. I added some upgrade queries into the timer so you wouldn't need to do this. Can you check to see if they ran?

LubricantJam commented:
> Yeah upgrade scripts will be needed. I am curious as to performance in larger checkpoint maps and also why 9 different fields for a velocity? I get you want some stats but is the entry velocity not good enough for checkpoints? Putting all that extra data seems slightly overkill especially for servers with thousands of completes on maps with higher checkpoints

You're asking the wrong person, I'm simply moving it over for people to use. 

I run this on a server with thousands upon thousands of completions, there is no issues there.

dPexxIAM commented:
I've run these changes for a year or so also. Performance does not get impacted too greatly.

LubricantJam commented:
@qawery-just-sad & @samatazz I have added a SQL file for you to import rather than depending on hard coded queries.

qawery-just-sad commented:
`L 04/25/2020 - 19:44:59: [SM] Exception reported: Language phrase "ShowVelocityCP" not found (arg 4)
L 04/25/2020 - 19:44:59: [SM] Blaming: surftimer.smx
L 04/25/2020 - 19:44:59: [SM] Call stack trace:
L 04/25/2020 - 19:44:59: [SM]   [0] VFormat
L 04/25/2020 - 19:44:59: [SM]   [1] Line 93, C:\Users\Qawery\Downloads\Surftimer-olokos-dev\addons\sourcemod\scripting\include\colorvariables.inc::CPrintToChat
L 04/25/2020 - 19:44:59: [SM]   [2] Line 4084, surftimer/misc.sp::Checkpoint
L 04/25/2020 - 19:44:59: [SM]   [3] Line 422, surftimer/surfzones.sp::StartTouch
L 04/25/2020 - 19:44:59: [SM]   [4] Line 236, surftimer/surfzones.sp::StartTouchTrigger`
I assume u still working on that

qawery-just-sad commented:
And yes queries are now correct

qawery-just-sad commented:
i see my mistake

LubricantJam commented:
There has only been 1 additional translation.

qawery-just-sad commented:
yes

qawery-just-sad commented:
`QSURF | Stage 1: 439 u/s [WR: +439 | PB: +439]
QSURF | Speed: 439 u/s [WR: +439 | PB: +439]`
I think you know what i mean

qawery-just-sad commented:
And this is just confusion 9000
QSURF | Start: 343 u/s [WR +343 | PB +343] | End: 86 u/s [WR +86 | PB +86]
Please include the stage number

LubricantJam commented:
> And this is just confusion 9000
> QSURF | Start: 343 u/s [WR +343 | PB +343] | End: 86 u/s [WR +86 | PB +86]
> Please include the stage number

I'm so lost with your spam of messages that I don't even know what you're on about.

qawery-just-sad commented:
1. It doubles the prestrafe message for stage one and start map 
2. So if i am correct it displays the prestrafe speed of last stage (for example stage 2) as well as speed when entering next (stage 3) stage? Why does it not display the prestrafe speed of the stage when on that stage?

samatazz commented:
when you say you have no issues running this on your server with thousands of completes, are you running you MYSQL on the same box as your server or externally? is it through a host or on a VPS?

These queries seem to be the way that the original ckSurf did them and im not sure how optimised these are for servers that are forced to use external MySQL but looking at the recent optimisations it seems completely likely these will be fine!

Another question is why save 6 different fields for speed? XY,XYZ, Z for each, Can you not use the pre-calculated speed that the hud uses and just pull that data on entry and exit? lowering it to only 2 fields for enter and leaving? also why do you need the enter speed at all for prestrafing? you run at the same speed consistently so wouldn't the leaving speed for stages/start be enough as its not like you can enter at different speeds unless its a telehopping map in which case leaving speed should be enough anyway as you're not gonna be strafing through the zones then either.

I know KSF save the speeds of when you leave/enter zones but they only use it once and I do think its an incredibly good stat that should be added to the timer but it should be a simple implementation.

LubricantJam commented:
> when you say you have no issues running this on your server with thousands of completes, are you running you MYSQL on the same box as your server or externally? is it through a host or on a VPS?
> 
> These queries seem to be the way that the original ckSurf did them and im not sure how optimised these are for servers that are forced to use external MySQL but looking at the recent optimisations it seems completely likely these will be fine!
> 
> Another question is why save 6 different fields for speed? XY,XYZ, Z for each, Can you not use the pre-calculated speed that the hud uses and just pull that data on entry and exit? lowering it to only 2 fields for enter and leaving? also why do you need the enter speed at all for prestrafing? you run at the same speed consistently so wouldn't the leaving speed for stages/start be enough as its not like you can enter at different speeds unless its a telehopping map in which case leaving speed should be enough anyway as you're not gonna be strafing through the zones then either.
> 
> I know KSF save the speeds of when you leave/enter zones but they only use it once and I do think its an incredibly good stat that should be added to the timer but it should be a simple implementation.

MySQL, different box in the same datacentre. You're asking me way too many questions when I said previously, I did not write this. You are best off asking @fluffyst not me.

samatazz commented:
Ah, Just figured you wrote it due to the pull request. Well in that case i think that some of this could be used but not sure we should approve the whole thing as it could be optimised better

LubricantJam commented:
> Ah, Just figured you wrote it due to the pull request. Well in that case i think that some of this could be used but not sure we should approve the whole thing as it could be optimised better

It's open to edits, feel free. I added the base and some additional changes that I felt were needed as I was going through the timer. The whole "field for checkpoint" structure definitely needs to go. The velocity stuff was just tied to it, so I added that too.

We could always strip it out for the time being if that's what people would like.

olokos commented:
I am a bit worried if this will increase the db disk space used multiple times aswell as saving all those coordinates, but @LubricantJam says his db is 8gb in size, can we get any other numbers from others running the same code?

Or at least db size of people who have large amounts of records in the db, but not running this code yet. As a means of comparison. 

I am also not very happy with saving 6fields for speed, but if no optimizations are provided after we compare db sizes before and after this code, I would totally merge this, while updating the Readme with the new sql update script, shortly after merging.

LubricantJam commented:
With this change in mind, my checkpoints table is 1.9 million rows at 148MB. This is with a total of nearly 100,000 players in the ck_playerrank table. That's an idea of scalability.

Bara commented on addons/sourcemod/scripting/surftimer/commands.sp:
```diff
@@ -111,22 +111,25 @@ void CreateCommands()
-	RegAdminCmd("sm_givevip", VIP_GiveVip, ADMFLAG_ROOT, "[surftimer] Give a player VIP");
-	RegAdminCmd("sm_removevip", VIP_RemoveVip, ADMFLAG_ROOT, "[surftimer] Remove a players VIP");
-	RegAdminCmd("sm_addcredits", VIP_GiveCredits, ADMFLAG_ROOT, "[surftimer] Give a player credits");
+	if (GetConVarBool(g_iVIPCommands))
```
I'm pretty sure, that you only get the default value instead of the custom value. I noticed this issue a few months/years ago on my test servers.

LubricantJam replied:
Nah works all fine, tested 👍

Bara commented on addons/sourcemod/scripting/surftimer/convars.sp:
```diff
@@ -163,9 +164,11 @@ void CreateConVars()
+	g_iAdminCountryTags = CreateConVar("ck_admin_country_tags", "0", "(1 / 0) Enable/Disable country tags for admins", FCVAR_NOTIFY, true, 0.0, true, 1.0);
 	g_replayBotDelay = CreateConVar("ck_replay_bot_delay", "10", "Delay in seconds after initial mapstart after the bots join the server", FCVAR_NOTIFY, true, 10.0);
+	g_iVIPCommands = CreateConVar("ck_vip_commands", "1", "(1 / 0) Enable/Disable all SurfTimer VIP commands.", FCVAR_NOTIFY, true, 0.0, true, 1.0);
+	g_iVIPCommands = FindConVar("ck_vip_commands");
```
And [FindConVar](https://sourcemod.dev/#/convars/function.FindConVar) is useless, because [CreateConVar](https://sourcemod.dev/#/convars/function.CreateConVar) already returns the handle.

LubricantJam replied:
You're right, however when I was testing the cvar wasn't picking up the value being set because commands are called right afterwards the convar creation.

Bara replied:
Moving the VIP Commands part into OnConfigsExecuted should work.

LubricantJam replied:
Done this now, cheers.

LubricantJam replied:
Don't think it actually worked, I'll investigate further.

LubricantJam replied:
False flag, community members fake reporting.

LubricantJam commented:
I will await for @sneak-it's reply before merging the PR.

fluffyst commented:
This method was in my "private" version of timer and it works fine, you can't calculate the values for each from just the normal speed, but this is what was used on kp, imk, etc

Please ALWAYS use an external server for your database!!!

This method also allows you to add in more stats too if someone ever decided to do it

sneak-it commented on addons/sourcemod/scripting/surftimer/commands.sp:
```diff
@@ -207,6 +185,34 @@ void CreateCommands()
+void CreateVIPCommands()
+{
+	// VIP Commands
+	RegAdminCmd("sm_fixbot", Admin_FixBot, g_VipFlag, "[surftimer] Toggles replay bots off and on");
+	RegAdminCmd("sm_fixbots", Admin_FixBot, g_VipFlag, "[surftimer] Toggles replay bots off and on");
+	RegAdminCmd("sm_fb", Admin_FixBot, g_VipFlag, "[surftimer] Toggles replay bots off and on");
```
If these commands aren't regulated by the VIP convar, they shouldn't be inside CreateVIPCommands function. I don't think they should be a VIP command, perhaps an admin command in which server owners could override to a different flag if they'd like.

LubricantJam replied:
Ay, glad someone else noticed this. Yeah, I didn't think they should be VIP Commands, I'll do exactly that and put them behind Admin Commands once I'm back at my computer. 

And yeah, I put them in there just to keep them all in one place, brilliant thanks for the feedback boss.

LubricantJam replied:
Added in latest commit.
